### ISSUE TITLE
Regression in 1.1.331 for `__new__` and/or `Self`

### ISSUE THREAD

**Issue by arjenzorgdoc** [microsoft/pyright#6118]
This minimal test case does not type check in v1.1.331:

```python
from typing_extensions import Self


class X(str):
    def __new__(cls, v: str) -> Self:
        if type(v) is cls:
            return v

        return super().__new__(cls, v)

i = X("a")

print(i is X("a")) # => False
print(i is X(i)) # => True
```

```shell
$ for e in pyright@1.1.{330..331}; do  echo =========== $e =============; npx $e test.py ; done
=========== pyright@1.1.330 =============
0 errors, 0 warnings, 0 informations 
=========== pyright@1.1.331 =============
/home/arjen/pyright-issue/test.py
  /home/arjen/pyright-issue/test.py:7:20 - error: Expression of type "str" cannot be assigned to return type "Self@X"
    "str" is incompatible with "X" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 
```

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by arjenzorgdoc** [comment]
It seems it doesn't work with unions:

test_union.py:
```python
from typing import Union
from typing_extensions import Self


class X(str):
    def __new__(cls, v: Union[str, int]) -> Self:
        if type(v) is cls:
            return v

        return super().__new__(cls, v)

i = X("a")

print(i is X("a")) # => False
print(i is X(i)) # => True
```

```
$ for e in pyright@1.1.{330..331} ~/pyright/packages/pyright; do  echo =========== $e =============; npx $e *; done
=========== pyright@1.1.330 =============
0 errors, 0 warnings, 0 informations 
=========== pyright@1.1.331 =============
/home/arjen/pyright-issue/test.py
  /home/arjen/pyright-issue/test.py:7:20 - error: Expression of type "str" cannot be assigned to return type "Self@X"
    "str" is incompatible with "X" (reportGeneralTypeIssues)
/home/arjen/pyright-issue/test_union.py
  /home/arjen/pyright-issue/test_union.py:8:20 - error: Expression of type "str | int" cannot be assigned to return type "Self@X"
    Type "str | int" cannot be assigned to type "X"
      "int" is incompatible with "X" (reportGeneralTypeIssues)
2 errors, 0 warnings, 0 informations 
=========== /home/arjen/pyright/packages/pyright =============
/home/arjen/pyright-issue/test_union.py
  /home/arjen/pyright-issue/test_union.py:8:20 - error: Expression of type "X | int" cannot be assigned to return type "Self@X"
    Type "X | int" cannot be assigned to type "X"
      "int" is incompatible with "X" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 
```

The last one is 495681027d6

**Comment by erictraut** [comment]
@arjenzorgdoc, I don't think it's safe to narrow the union in your example.

Consider the following code. What would you expect `reveal_type` to reveal?

```python
class A: ...
class B: ...

def func1(x: A, t: type[B]):
    if type(x) is t:
        reveal_type(x) # What do you think the narrowed type should be?
```

I'm guessing that you said `Never`, since there's clearly no overlap between type `A` and `B`. However, consider the following:

```python
def func1(x: A, t: type[B]):
    if type(x) is t:
        return True
    return False

class C(A, B): ...

print(func1(C(), C))
```

So `Never` isn't the correct answer here. The correct answer is to not narrow the type in this case. For the same reason, it's not safe to assume in your example above that `int` is eliminated by the type guard condition.

**Comment by erictraut** [comment]
This is included in pyright 1.1.332, which I just published. It will also be included in a future release of pylance.

**Comment by arjenzorgdoc** [comment]
I don't know how to properly type my `__new__` method now. It used to work in 1.1.330. 

> @arjenzorgdoc, I don't think it's safe to narrow the union in your example.

I don't think I understand this. I cannot think of any way the type assertion could fail at runtime. It will always return a "`Self`" (`X` or the proper subclass).

> [...] it's not safe to assume in your example above that int is eliminated by the type guard condition.

It's only not eliminated if `Self` is a subclass of `int`. That still implies `isinstance(v, Self)`. (you are right that it doesn't imply `not isinstance(v, int)`)

> 
> Consider the following code. What would you expect `reveal_type` to reveal?
> 
> ```python
> class A: ...
> class B: ...
> 
> def func1(x: A, t: type[B]):
>     if type(x) is t:
>         reveal_type(x) # What do you think the narrowed type should be?
> ```

Maybe when using `type(x) is B` it should always narrow to exactly `B`? So `B`? (or something like B@func1)?

Typescript says: `A & B` but that satisfies `B`:

```typescript
class A {};

function func1<B, TB extends new () => B>(a: A, t: TB): B {
    if (a instanceof t) {
        return a;
        //     ^?  # (parameter) a: A & B
    }

    return new t();
}
```

If I try to type my `__new__` function in typescript it also works:

```typescript
class X extends String {}

function __new__<Int, Self extends X, TSelf extends new (a: string) => Self>(cls: TSelf, a: Int | String) {
  if (a instanceof cls) {
    return a;
    //     ^? # (parameter) a: Self extends X
  }

  return new cls("b");
} 

const a = new X("a");
const b = __new__(X, a);
//    ^? # const b: X
console.log(a === b)
```

**Comment by arjenzorgdoc** [comment]
Wait now I'm really confused:

```python
class A: ...
class B: ...

def func1(x: A, t: type[B]):
    if type(x) is t:
        reveal_type(x)
 # .../x.py:6:21 - information: Type of "x" is "A"

    if type(x) is t and isinstance(x, t):
        reveal_type(x)
#  .../x.py:9:21 - information: Type of "x" is "<subclass of A and B>"
```


I thought that `type(x) is t` implies `isinstance(x, t)`.

I get rid of the error by adding an `isinstance`:

```python
from typing import Union
from typing_extensions import Self


class X(str):
    def __new__(cls, v: Union[str, int]) -> Self:
        if type(v) is cls and isinstance(v, cls):
            return v

        return super().__new__(cls, v)
```

That doesn't make sense to me.
